### PR TITLE
Implement ExtensionIndex

### PR DESCRIPTION
**Pull request by jbrockmendel**
So far this just puts `take` in it, but there are a handful of other methods we'll be able to move from our existing EA Indexes.

### CODE REVIEW

File: pandas/core/indexes/category.py [hunk 1]
```diff
@@ -722,20 +721,9 @@ def _convert_arr_indexer(self, keyarr):
...
-            self._data,
-            indices,
-            allow_fill=allow_fill,
-            fill_value=fill_value,
-            na_value=self._data.dtype.na_value,
-        )
-        return self._shallow_copy(taken)
-
-    take_nd = take
+    def take_nd(self, *args, **kwargs):
```
Comment (jreback) [inline pandas/core/indexes/category.py]: didn't you deprecate this?

Comment (jbrockmendel) [reply pandas/core/indexes/category.py]: #30702, will need to be rebased

**Comment by jbrockmendel** [comment]
rebased+green

**Comment by jreback** [comment]
k thanks

**Pull request merged**